### PR TITLE
Sequential signing, tails & veracrypt support, qr-upgrades, et al.

### PR DESCRIPTION
**Pull request by ak1n**
I wanted to implement glacier but did not want paper keys or co-localization of keys. I thus made modifications including:
* sequential/re-signing support
* tails-veracrypt support
* minor changes:
  * increment qr filenames rather than overwriting
  * generate qr via glacier script argument
  * verbose mode
  * mBTC fee display
  * single safety confirmation

More detailed modification and testing information on master README at: https://github.com/ak1n/GlacierProtocol

I would appreciate any feedback/thoughts/suggestions. I am hoping to get more formal auditing to ensure there aren't new egregious code or protocol errors, but wanted to post to this main repo first. I understand that not everyone might want all the code changes above so separated the modifications into discrete commits/branches in the fork. Thanks for any feedback! -ak1n

### CODE REVIEW

File: glacierscript.py [hunk 1]
```diff
@@ -3,15 +3,15 @@
 ################################################################################################
 #
 # GlacierScript:  Part of the Glacier Protocol (http://glacierprotocol.org)
-#
```
Comment (bitcoinhodler) [inline glacierscript.py]: Commingling trivial changes like this with substantive changes in a single commit makes review unnecessarily difficult. Best to make a separate commit with only the trivial changes next time.

File: glacierscript.py [hunk 2]
```diff
@@ -341,6 +342,18 @@ def get_utxos(tx, address):
...
     return utxos
 
+def verbose(content):
+    # if verbose mode enabled, print content
+    if VERBOSE_MODE:
+        print content
+
+def bitcoin_cli_call(cmd,args):
+    full_cmd = "{0}{1} {2}".format(bitcoin_cli,cmd,args)
+    # note glacier has a space after bitcoind call in "bitcoin_cli" variable
```
Comment (bitcoinhodler) [inline glacierscript.py]: Why not remove that space now, and move it to full_cmd? It was only there to make the usage of it a little cleaner, but that's gone now.

File: glacierscript.py (hunk 1 above)
Comment (ak1n) [reply glacierscript.py]: Thanks for the response. Understood re pull request separation - I am a novice at this as usually hacking only my own code. Should I re-submit my individual branches (sequential-sign, tails, base-upgrades)?

File: glacierscript.py [hunk 3]
```diff
@@ -317,11 +319,10 @@ def addmultisigaddress(m, addresses_or_pubkeys, address_type='p2sh-segwit'):
     require_minimum_bitcoind_version(160000) # addmultisigaddress API changed in v0.16.0
     address_string = json.dumps(addresses_or_pubkeys)
     argstring = "{0} '{1}' '' '{2}'".format(m, address_string, address_type)
-    return json.loads(subprocess.check_output(
-        bitcoin_cli + "addmultisigaddress {0}".format(argstring), shell=True))
+    return json.loads(bitcoin_cli_call("addmultisigaddress",argstring))
```
Comment (bitcoinhodler) [inline glacierscript.py]: Code formatting: please put a space after commas

File: glacierscript.py (hunk 1 above)
Comment (bitcoinhodler) [reply glacierscript.py]: Don't bother re-submitting...they are reasonably-enough segregated already.

File: glacierscript.py [hunk 4]
```diff
@@ -40,6 +40,8 @@
 
 SATOSHI_PLACES = Decimal("0.00000001")
 
+VERBOSE_MODE = 0
```
Comment (bitcoinhodler) [inline glacierscript.py]: In python, all-caps is a convention for global constants (see [Google's python style guide](https://google.github.io/styleguide/pyguide.html). In this case, this is not constant, as a cmdline option could change it.

File: glacierscript.py (hunk 2 above)
Comment (ak1n) [reply glacierscript.py]: Will do...I'll incorporate into my base-upgrades branch. Thank you!

File: glacierscript.py [hunk 5]
```diff
@@ -537,7 +541,17 @@ def safety_checklist():
         "Are smartphones and all other nearby devices turned off and in a Faraday bag?"]
 
     for check in checks:
-        answer = raw_input(check + " (y/n)?")
```
[the remark this answers is not in the document]
Comment (ak1n) [reply glacierscript.py]: I was focused on testing with real coins to get closer to what I'll eventually be doing, so I neglected this. I will review this and re-code - thank you for drawing my attention to this!

File: glacierscript.py [hunk 6]
```diff
@@ -493,20 +493,35 @@ def write_and_verify_qr_code(name, filename, data):
...
+    filename: <string> filename for storing the QR code. note that ".png" is added in function to make incrementing easier
     data: <string> the data to be encoded
     """
 
-    subprocess.call("qrencode -o {0} {1}".format(filename, data), shell=True)
+    # note: it would probably be better to write qr codes to RAM disk or secure volume. while shouldn't contain keys, may increase security by decreasing chance of unencrypted transaction history being discovered
+    QR_SUBDIR = "qrcodes"
+    QR_SUFFIX = ".png"
+    script_root = os.path.dirname(os.path.abspath(__file__))
+    QR_DIRPATH = script_root + "/" + QR_SUBDIR
```
Comment (bitcoinhodler) [inline glacierscript.py]: This moves the qr codes from their previous location, which means the Glacier doc needs to change to tell users where to find the file.

File: glacierscript.py [hunk 7]
```diff
@@ -493,20 +493,35 @@ def write_and_verify_qr_code(name, filename, data):
     Write a QR code and then read it back to try and detect any tricksy malware tampering with it.
 
     name: <string> short description of the data
-    filename: <string> filename for storing the QR code
+    filename: <string> filename for storing the QR code. note that ".png" is added in function to make incrementing easier
     data: <string> the data to be encoded
     """
 
-    subprocess.call("qrencode -o {0} {1}".format(filename, data), shell=True)
+    # note: it would probably be better to write qr codes to RAM disk or secure volume. while shouldn't contain keys, may increase security by decreasing chance of unencrypted transaction history being discovered
```
Comment (bitcoinhodler) [inline glacierscript.py]: Everything on the quarantined laptops is run from RAM disk already.

File: glacierscript.py [hunk 8]
```diff
@@ -493,20 +493,35 @@ def write_and_verify_qr_code(name, filename, data):
...
+    QR_SUFFIX = ".png"
+    script_root = os.path.dirname(os.path.abspath(__file__))
+    QR_DIRPATH = script_root + "/" + QR_SUBDIR
+    if not os.path.isdir(QR_DIRPATH):
+        os.mkdir(QR_DIRPATH)
+    QR_PATH = QR_DIRPATH + "/" + filename + QR_SUFFIX
+    if os.path.exists(QR_PATH):
+        #print "QR exists at: {0}".format(QR_PATH)
+        i = 2
+        while os.path.exists(QR_DIRPATH + "/" + filename + str(i) + QR_SUFFIX):
```
Comment (bitcoinhodler) [inline glacierscript.py]: Ugly code duplication here -- construction of QR_PATH appears three separate times. That must be in one place (principle of DRY: Don't Repeat Yourself).

File: glacierscript.py [hunk 9]
```diff
@@ -840,6 +855,7 @@ def withdraw_interactive():
     parser.add_argument(
         "-n", type=int, help="Number of total keys required in an m-of-n multisig address creation (default m-of-n = 1-of-2)", default=2)
     parser.add_argument('--testnet', type=int, help=argparse.SUPPRESS)
+    parser.add_argument("-q", "--qrdata", help="Data to be encoded into qr-code")
```
Comment (bitcoinhodler) [inline glacierscript.py]: What is the purpose and use case of this flag? And of this entire commit? Creating multiple withdrawals in one sitting?

File: glacierscript.py [hunk 10]
```diff
@@ -797,7 +800,7 @@ def withdraw_interactive():
                 print "{0} BTC going back to cold storage address {1}".format(value, address)
             else:
                 print "{0} BTC going to destination address {1}".format(value, address)
-        print "Fee amount: {0}".format(fee)
+        print "Fee amount: {0} btc ({1} mbtc)".format(fee,btc_to_mbtc(fee))
```
Comment (bitcoinhodler) [inline glacierscript.py]: DRY this up...two places where fee amount display is computed, and they are inconsistent (bitcoin/btc, mbtc/mBTC).

File: glacierscript.py [hunk 11]
```diff
@@ -88,6 +88,9 @@ def btc_to_satoshi(btc):
     value = btc * 100000000
     return int(value)
 
+def btc_to_mbtc(btc):
+    mbtc = Decimal(btc)*1000
+    return mbtc.quantize(SATOSHI_PLACES)
```
Comment (bitcoinhodler) [inline glacierscript.py]: Won't this show mBTC with 8 decimal places? Which is a little confusing as mBTC only divides to 5 decimal places.

File: glacierscript.py [hunk 12]
```diff
@@ -493,20 +493,35 @@ def write_and_verify_qr_code(name, filename, data):
...
 
-    subprocess.call("qrencode -o {0} {1}".format(filename, data), shell=True)
+    # note: it would probably be better to write qr codes to RAM disk or secure volume. while shouldn't contain keys, may increase security by decreasing chance of unencrypted transaction history being discovered
+    QR_SUBDIR = "qrcodes"
+    QR_SUFFIX = ".png"
+    script_root = os.path.dirname(os.path.abspath(__file__))
+    QR_DIRPATH = script_root + "/" + QR_SUBDIR
+    if not os.path.isdir(QR_DIRPATH):
+        os.mkdir(QR_DIRPATH)
+    QR_PATH = QR_DIRPATH + "/" + filename + QR_SUFFIX
```
Comment (bitcoinhodler) [inline glacierscript.py]: QR_PATH and QR_DIRPATH are not constants; shouldn't be all caps

File: glacierscript.py [hunk 13]
```diff
@@ -706,21 +709,84 @@ def withdraw_interactive():
...
-        addresses[source_address] = 0
 
-        redeem_script = raw_input("\nRedemption script for source cold storage address: ")
+        if RE_SIGN_MODE is not 1:
+            redeem_script = raw_input("\nRedemption script for source cold storage address: ")
+            dest_address = raw_input("\nDestination address: ")
+            num_tx = int(raw_input("\nHow many unspent transactions will you be using for this withdrawal? "))
+        else:
+            # begin main re-sign code block
+            print "\nPlease paste the partially-signed raw transaction (hexadecimal format) with unspent outputs at the source address"
```
Comment (bitcoinhodler) [inline glacierscript.py]: Ugly code duplication here

File: glacierscript.py [hunk 14]
```diff
@@ -706,21 +709,84 @@ def withdraw_interactive():
...
+            print "(If the transaction data is over ~4000 characters long, you _must_ use a file.):"
 
-        dest_address = raw_input("\nDestination address: ")
-        addresses[dest_address] = 0
+            part_signed_hex_tx = raw_input()
+            if os.path.isfile(part_signed_hex_tx):
+                part_signed_hex_tx = open(part_signed_hex_tx).read().strip()
+
+            part_signed_tx = json.loads(bitcoin_cli_call("decoderawtransaction",part_signed_hex_tx))
+            redeem_script=part_signed_tx["vin"][0]["txinwitness"][-1]
```
Comment (bitcoinhodler) [inline glacierscript.py]: Will this work if the Glacier cold storage address is an older, pre-segwit address? It's important that Glacier continue to fully work for all previously-created addresses.

File: glacierscript.py [hunk 15]
```diff
@@ -706,21 +709,84 @@ def withdraw_interactive():
...
+            print "\n    number of input transactions: {0}".format(num_tx)
+            print "\n    change amount: {0}".format(change_amount)
+            print "\n    withdrawal amount: {0}".format(withdrawal_amount)
+
+            print "\n\nplease confirm whether above data is correct before proceeding to input additional data for transaction re-sign"
+            confirm = yes_no_interactive()
+            if not confirm:
+                print "auto parsed data from transaction incorrect so aborting"
+                sys.exit()
+            # end main re-sign code block
```
Comment (bitcoinhodler) [inline glacierscript.py]: This whole block should be pulled out into separate function(s).

File: glacierscript.py [hunk 16]
```diff
@@ -816,11 +886,25 @@ def withdraw_interactive():
     #### Calculate Transaction ####
     print "\nCalculating transaction...\n"
 
-    unsigned_tx = create_unsigned_transaction(
-        source_address, addresses, redeem_script, txs)
+    if RE_SIGN_MODE is not 1:
```
Comment (bitcoinhodler) [inline glacierscript.py]: I'm no Python expert but your use of Booleans like this doesn't seem like best practice. Wouldn't it be better to set the variable to True or False and test using `if RE_SIGN_MODE:` and `if not RE_SIGN_MODE:`

File: glacierscript.py [hunk 17]
```diff
@@ -816,11 +886,25 @@ def withdraw_interactive():
...
+            source_address, addresses, redeem_script, input_txs)
 
-    signed_tx = sign_transaction(source_address, keys,
-                                 redeem_script, unsigned_tx, txs)
+        signed_tx = sign_transaction(source_address, keys,
+                                     redeem_script, unsigned_tx, input_txs)
+    else:
+        re_sign_input_txs = []
+        txid = tx["txid"]
+        for utxo in utxos:
```
Comment (bitcoinhodler) [inline glacierscript.py]: This looks like nasty code duplication

File: glacierscript.py [hunk 18]
```diff
@@ -884,5 +968,10 @@ def withdraw_interactive():
     if args.program == "create-withdrawal-data":
         withdraw_interactive()
 
+    if args.program == "sign-transaction":
+        # Sign an existing transaction (i.e. add a signature to partially signed tx) to withdaw funds from cold storage
+        RE_SIGN_MODE = 1
+        withdraw_interactive()
```
Comment (bitcoinhodler) [inline glacierscript.py]: Related to my comment above (about extracting that block into a separate function), perhaps it would be cleaner to have two methods, e.g withdraw_new() and withdraw_continue(), implemented separately, with the common parts pulled out into functions.

File: glacierscript.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bitcoinhodler) [reply glacierscript.py]: We need a contributor's guide (see #27) to explain this to people.

File: glacierscript.py (hunk 1 above)
Comment (bitcoinhodler) [reply glacierscript.py]: I think I misunderstood you here...I meant that the commits are segregated. I understand now you are asking about separate PRs, and yes I think separate PRs would be good.

File: glacierscript.py (hunk 9 above)
Comment (ak1n) [reply glacierscript.py]: The multi withdrawals in one session is why I wrote the qr suffix increment. I put in this qr-code -q business to handle getting secure data out of q environment in a secure-ish way. This is important/needed for me because I don't want to use paper keys and plan on using veracrypted volumes/data/keys. An example of data to export would be the cold storage public address (post the initial script run). People using paper keys have it on their paper, but I won't and might need to re-acquire it for a new location. Then there is "why not just use qrencode"? I like the feedback checking the already-written qr function so that's why I introduced the CLI argument rather than planning on using qrencode independently (I could be wrong about the advantages here though). Thanks again for reviewing all of this.

File: glacierscript.py (hunk 2 above)
Comment (ak1n) [reply glacierscript.py]: In my development branch I've taken out the space in the bitcoin cli & related arg variables and moved the space into this bitcoin_cli_call function. To not interfere with other bitcoin-cli calls (these are handled differently re e.g. shell options), I've added the space to these other calls. Thanks!

File: glacierscript.py (hunk 6 above)
Comment (ak1n) [reply glacierscript.py]: While I think being able to move to different path might have future value (e.g. archiving the QRs in a secure volume in q-environment), to simplify at this point in time I've reverted the destination directory back to glacier root. Thanks!

File: glacierscript.py (hunk 7 above)
Comment (ak1n) [reply glacierscript.py]: Thank you - I've removed this erroneous comment.

File: glacierscript.py (hunk 9 above)
Comment (bitcoinhodler) [reply glacierscript.py]: Hmm...still not sure I understand (the cold storage address is already QR-encoded as part of the create-deposit-data, why do you need to QR-encode it again later?), but it sounds like this is a niche use case that doesn't really belong in this global repo. Better not to have code in GlacierScript that is not documented and has no use case except for your special situation.

**Comment by bitcoinhodler** [comment]
@ak1n please close this PR (if you can) since we agree it should not be merged as-is.

**Comment by ak1n** [comment]
@bitcoinhodler thanks for the help! Hope to re-submit soon integrating your input.

File: glacierscript.py (hunk 10 above)
Comment (ak1n) [reply glacierscript.py]: I'm not sure exactly what you mean by this? I put this in because I just wanted an additional display in mBTC (in parentheses) to ensure I wasn't counting zeroes wrong in BTC. The only calculation/computation should be just multiplication by 1000 (not an actual full fee recalculation). Any clarification would be appreciated!

File: glacierscript.py (hunk 9 above)
Comment (ak1n) [reply glacierscript.py]: I think that if more people elect to not use paper storage it will be less of a niche case, under the assumption that diligent QRs are not necessarily being kept (easier to store such information in keepass inside a veracrypt volume). I will go through the doc revisions once I have my code additions/proposals in respectable form. But I will respect your opinion and segregate this out. Thanks!

File: glacierscript.py (hunk 10 above)
Comment (bitcoinhodler) [reply glacierscript.py]: Lines 477 and 803 (as of a0c9c6019a6ae5e122073f8a3bf466f59fe72d1e) both do the same thing, with subtle inconsistencies.

`{0} bitcoin = {1} mBTC`
`{0} btc ({1} mbtc)`

This formatting should be done in one place -- and might as well use it everywhere a BTC amount is displayed.

This will break the developer tests, so please include an update of those in the same commit.

Comment (ak1n) [reply glacierscript.py]: Understood - thank you for the clarification.

File: glacierscript.py (hunk 9 above)
Comment (bitcoinhodler) [reply glacierscript.py]: Using electronic storage for keys is a major change to Glacier and its security model. IMO that doesn't belong in this primary Glacier repo, even as an option. (See principle 5 in the design document: Offer Limited Choice.)

I'd recommend you do your own fork of Glacier to support electronic storage of keys.

Comment (ak1n) [reply glacierscript.py]: Understood, will do - thanks for the response!

File: glacierscript.py (hunk 16 above)
Comment (ak1n) [reply glacierscript.py]: I'm no Python expert either...I initially set out to use True/False, but after discovering that in Python 2 (which glacier is written for) these booleans are mutable I switched to 0 and 1 (though there should be low probability of mutability in practice). https://stackoverflow.com/questions/2764017/is-false-0-and-true-1-in-python-an-implementation-detail-or-is-it-guarante

I don't know what best practice is...if it is using True/False I can obviously clean things up as booleans for consolidation/readability.

File: glacierscript.py (hunk 14 above)
Comment (bitcoinhodler) [reply glacierscript.py]: Glacier being designed for non-technical users, many people using it might have no idea if their cold storage address is segwit or not.

In the developer tests, there is a withdrawal using non-segwit. You could use those keys to test the non-segwit sequential signing. I suspect the code required to handle it will be simple.

File: glacierscript.py (hunk 16 above)
Comment (bitcoinhodler) [reply glacierscript.py]: That is pretty crazy of Python 2 but as we're not going to be reassigning True/False to anything other than 1/0 respectively, I see no issue with using them, and it makes it clearer that the variable is a boolean and not an integer.

Even if using 0/1, testing the value can still use `if RE_SIGN_MODE:` without the `is 1` noise.
